src/core/SamplePreviewPlayHandle.cpp

```diff
@@ -0,0 +1,142 @@
+/*
+ * SamplePreviewPlayHandle.cpp
+ *
+ * Copyright (c) 2025 Sotonye Atemie <sakertooth@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SamplePreviewPlayHandle.h"
+
+#include "AudioBusHandle.h"
+#include "AudioEngine.h"
+#include "Engine.h"
+#include "SampleFrame.h"
+#include "ThreadPool.h"
+
+namespace lmms {
+SamplePreviewPlayHandle::SamplePreviewPlayHandle(const std::filesystem::path& path)
+	: PlayHandle(Type::SamplePlayHandle)
+{
+#ifdef LMMS_BUILD_WIN32
+	m_sndfile = sf_wchar_open(path.c_str(), SFM_READ, &m_sfinfo);
+#else
+	m_sndfile = sf_open(path.c_str(), SFM_READ, &m_sfinfo);
+#endif
+
+	if (m_sndfile == nullptr || m_sfinfo.channels == 0 || m_sfinfo.frames == 0)
+	{
+		throw std::runtime_error{"Failed to create sample preview stream"};
+	}
```

Comment:
```suggestion
	if (!m_sndfile)
	{
		throw std::runtime_error{"Failed to create sample preview stream: " + std::string{sf_strerror()}};
	}

	if (m_sfinfo.channels == 0 || m_sfinfo.frames == 0)
	{
		sf_close(m_sndfile);
		throw std::runtime_error{"Failed to create sample preview stream: Invalid SF_INFO"};
	}
```
Better diagnostics + handle cleanup when constructor fails because the destructor won't be called.

Reply:
Huh? Throwing in the constructor doesn't call the destructor? Really?

Reply:
Nope, the object isn't fully constructed if an exception is thrown from the constructor, and calling the destructor on the partially-constructed object could be dangerous
https://godbolt.org/z/nPWToMsr6